### ISSUE TITLE
unix visibility control missing

### ISSUE THREAD

**Issue by skaller** [eao197/so-5-5#17]
Looking at h/declspec.hpp it seems visibility control is only implemented for Windows.
I use this:

```
#if FLX_WIN32
  #if defined(FLX_STATIC_LINK)
    #define FLX_EXPORT
    #define FLX_IMPORT
  #else
    #define FLX_EXPORT __declspec(dllexport)
    #define FLX_IMPORT __declspec(dllimport)
  #endif
#else
  // All modules on Unix are compiled with -fvisibility=hidden
  // All API symbols get visibility default
  // whether or not we're static linking or dynamic linking (with -fPIC)
  #define FLX_EXPORT __attribute__((visibility("default"))) 
  #define FLX_IMPORT __attribute__((visibility("default"))) 
#endif
```

In particular note, you should compile with `-fvisibility=hidden`, and set the same visibility macros as shown for Unix as above. For you, these are `SO_5_TYPE` and `SO_5_FUNC`. You may need to fix your build system to ensure the default visibility is hidden.

As a comment, hidden visibility is not currently possible with static libraries. However shared libraries should always use two level namespace control. C++ Modules_TS may improve visibility control for static linkage. Its not clear if it will get rid of the need for macros like those above.

**Comment by eao197** [comment]
Thank you. We will investigate this and fix in the next update/release.

**Comment by skaller** [comment]
Thanks, but I'm building from the repository so it would be nice to know when the source is fixed so I can get my build tools to build it properly. Actually I need to create a Felix binding and I'm using nm 

```
nm -jgU installed_files/lib/libso.5.5.24.dylib | c++filt
```

to find the symbols initially, at the moment I get a whole heap of useless implementation details. The other method requires parsing the #include files, either automatically, manually, or both, and/or trying to parse the documentation files (which is even worse).  For the `nm` method to work, the non public API must be invisible.

BTW: a patch to the declspec.hpp file still won't fix the problem with template instantiations. Unfortunately I don't know how to fix that. not sure if manually instantiated templates allow the visibility specification. This is a design fault in the C++ Standard. It fails to account for shared libraries. The abstract machine is also deficient in failing to allow duplication of symbols which are inevitable with dynamic loading.

**Comment by eao197** [comment]
> it would be nice to know when the source is fixed

I let you know when fixes will be merged into the main development branch.

**Comment by eao197** [comment]
You can try the current version from the master branch.
It works on Linux, but some tests are failed on MacOS in TravisCI builds. I don't know why -- sometimes TravisCI shows very strange results, I suppose it is because of weak VM with MacOS.
I don't have access to actual hardware with MacOS on it.

**Comment by skaller** [comment]
It builds but seems no different. There are still a lot of symbols visible in "impl" namespaces. Maybe its correct.
[tmp.log](https://github.com/eao197/so-5-5/files/2751386/tmp.log)

**Comment by skaller** [comment]
I comment that if everything is exported .. it more or less just works like static linkage. In other words basic linkage operation still works, its just a bit slow. Its more of an issue in C, but C++ has namespaces to help.

However more advanced systems will have problems. Consider a client U has two libraries A and B, which both happen to use so_5 .. but they use *different versions*.  U doesn't know that A and B happen to use so_5, doesn't care, and can't see the symbols (with dynamic linkage). A and B each load *different* versions of so_5. With two level symbol linkage, there's no problem.

The problem for me is creating a binding. I built one for GNU Scientific Library which has something like 4 thousand functions in it. Felix bindings are lightweight because Felix is C++, but they still have to exist. It really helps if I can generate the bindings mechanically with a script. Usually i have to hand edit the result afterwards. 

~/so-5-5/cmake_build>wc tmp.log
    4004   33522 1462108 tmp.log

says so_5 has 4 thousand symbols defined. Is the API really that big?

**Comment by eao197** [comment]
> It builds but seems no different.

It's strange. On my Linux machine for 5.5.24 (without -fvisibility):
```{.sh}
nm -C -D target/gcc_5_5_0__x86_64_linux_gnu/release/libso.5.5.24.so | wc -l
3469
```
and for 5.5.24.1 (with -fvisibility and -fvisibility-inlines-hidden):
```{.sh}
nm -C -D target/gcc_5_5_0__x86_64_linux_gnu/release/libso.5.5.24.1.so | wc -l
991
```
CMake build produces even less numbers for 5.5.24.1:
```{.sh}
nm -C -D target/lib/libso.5.5.24.1.so | wc -l
983
```
Did you check the version from master branch at commit [2816b1f4c50e74b103779f1293552084c5a89c85](https://github.com/eao197/so-5-5/commit/2816b1f4c50e74b103779f1293552084c5a89c85)?

**Comment by skaller** [comment]
AH. I forgot to install. SORRY! With uninstalled files:

```
~/so-5-5/cmake_build>nm -jgU so_5/libso.5.5.24.dylib  | wc
    4004    4004  653299
~/so-5-5/cmake_build>nm -jgU so_5/libso.5.5.24.1.dylib  | wc
     520     520   36080
```

-j means "just symbols"
-g means "only global symbols"
-U means "don't display undefined symbols"

The output includes typeinfo and vtables. It looks very nice now. Thanks!

**Comment by eao197** [comment]
Can I ask you for help?

With -fvisibility some tests fails in TravisCI builds. But TravisCI builds sometimes surprise me with some unexpected errors. So I don't trust those builds much.

Can you build so-5.5.24.1 with tests enabled and then run those tests on your machine? It is like:
```{.sh}
cmake -DCMAKE_INSTALL_PREFIX=target -DBUILD_TESTS=ON -DCMAKE_BUILD_TYPE=Release ../dev
cmake --build . --config Release
cmake --build . --config Release --target test
```

**Comment by skaller** [comment]
I did build with tests enabled. I will try ```make test``` is that the same?  It is running the tests...

```
  9/307 Test   #9: _unit.test.environment.stop_guards.parallel_remove ..........................   Passed    0.06 sec
        Start  10: _unit.test.environment.stop_guards.parallel_remove_50k
 10/307 Test  #10: _unit.test.environment.stop_guards.parallel_remove_50k ......................***Exception: Other 30.01 sec
        Start  11: _unit.test.environment.stop_guards.simple
 11/307 Test  #11: _unit.test.environment.stop_guards.simple ...................................   Passed    0.01 sec

262/307 Test #262: _unit.test.internal_stats.simple_coop_count .................................***Exception: Other 20.01 sec
        Start 263: _unit.test.internal_stats.simple_named_mbox_count
263/307 Test #263: _unit.test.internal_stats.simple_named_mbox_count ...........................***Exception: Other 20.01 sec
        Start 264: _unit.test.internal_stats.simple_timer_thread
264/307 Test #264: _unit.test.internal_stats.simple_timer_thread ...............................***Exception: Other 20.01 sec
        Start 265: _unit.test.internal_stats.simple_work_thread_activity
265/307 Test #265: _unit.test.internal_stats.simple_work_thread_activity .......................***Exception: Other 20.01 sec


277/307 Test #277: _unit.test.env_infrastructure.simple_mtsafe_st.stats_on .....................***Failed    2.22 sec
        Start 278: _unit.test.env_infrastructure.simple_mtsafe_st.stats_coop_count
278/307 Test #278: _unit.test.env_infrastructure.simple_mtsafe_st.stats_coop_count .............***Exception: Other 20.01 sec
        Start 279: _unit.test.env_infrastructure.simple_mtsafe_st.stats_wt_activity
279/307 Test #279: _unit.test.env_infrastructure.simple_mtsafe_st.stats_wt_activity ............***Exception: Other 20.01 sec
        Start 280: _unit.test.env_infrastructure.simple_not_mtsafe_st.empty_init_fn


289/307 Test #289: _unit.test.env_infrastructure.simple_not_mtsafe_st.stats_on .................***Failed    2.22 sec
        Start 290: _unit.test.env_infrastructure.simple_not_mtsafe_st.stats_coop_count
290/307 Test #290: _unit.test.env_infrastructure.simple_not_mtsafe_st.stats_coop_count .........***Exception: Other 20.01 sec
        Start 291: _unit.test.env_infrastructure.simple_not_mtsafe_st.stats_wt_activity
291/307 Test #291: _unit.test.env_infrastructure.simple_not_mtsafe_st.stats_wt_activity ........***Exception: Other 20.01 sec



96% tests passed, 11 tests failed out of 307

Total Test time (real) = 317.98 sec

The following tests FAILED:
	 10 - _unit.test.environment.stop_guards.parallel_remove_50k (OTHER_FAULT)
	262 - _unit.test.internal_stats.simple_coop_count (OTHER_FAULT)
	263 - _unit.test.internal_stats.simple_named_mbox_count (OTHER_FAULT)
	264 - _unit.test.internal_stats.simple_timer_thread (OTHER_FAULT)
	265 - _unit.test.internal_stats.simple_work_thread_activity (OTHER_FAULT)
	277 - _unit.test.env_infrastructure.simple_mtsafe_st.stats_on (Failed)
	278 - _unit.test.env_infrastructure.simple_mtsafe_st.stats_coop_count (OTHER_FAULT)
	279 - _unit.test.env_infrastructure.simple_mtsafe_st.stats_wt_activity (OTHER_FAULT)
	289 - _unit.test.env_infrastructure.simple_not_mtsafe_st.stats_on (Failed)
	290 - _unit.test.env_infrastructure.simple_not_mtsafe_st.stats_coop_count (OTHER_FAULT)
	291 - _unit.test.env_infrastructure.simple_not_mtsafe_st.stats_wt_activity (OTHER_FAULT)
Errors while running CTest
make: *** [test] Error 8

```

**Comment by eao197** [comment]
Thanks!

I receive the same results on TravisCI builds. There are no those errors without -fvisibility flag.

It is hard to say what the reason. Maybe some hidden bug in SObjectizer. Maybe -fvisibility breaks something on MacOS (but not on Linux). I will try to investigate it.

**Comment by skaller** [comment]
That's with High Sierra standard clang:

```
~/so-5-5/cmake_build>clang --version
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
I also have

```
/usr/local/Cellar/llvm/7.0.0/bin/clang++ --version
clang version 7.0.0 (tags/RELEASE_700/final)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /usr/local/Cellar/llvm/7.0.0/bin
```

and

```~/so-5-5/cmake_build>g++ --version
g++ (Homebrew GCC 8.2.0) 8.2.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by eao197** [comment]
Can you try to run tests with GCC? I mean build with GCC and run tests then.

**Comment by skaller** [comment]
Can I ask about use of dynamic cast and exception handling? Are these involved in the tests at all?

There is a problem with these on all compilers because the C++ language is broken with respect to dynamic linkage and the compiler implementors (at least gnu and clang) compounded the error by using a wrong design for exception handling and RTTI.

Roughly the fault is that they use a pointer to RTTI related stuff in EH and dynamic_cast, but this is wrong design, it cannot work. In particular, with two binary objects, be they shared libraries or a shared library and an executable, since these binaries are separately linked, if the use the same C++ class type with virtual methods, they must both get separate vtables. Then if an exception is thrown across the boundary, or a dynamic cast is done, it fails, because they design requires the pointer to be unique, and it isn't.

To fix this wrong design a special kind of linker section was invented which is supposed to merge certain symbols. This make work with load time dynamic linkage but can't possibly work with run time dynamic linkage as far as I can see. (the only correct method requires maintaining an equivalence class of each symbol .. it can be done by using the mangled name as a canonical representative because that is unique even if the actual data are not)

Now, here the real crux of the problem: templates. How does one handle template specialisations? Again there is special inker handling to avoid duplication. However for functions and stuff duplication just wastes space, it doesn't matter unless some idiot programmer actually relies on the wrong Standard and tries comparing function pointers. BUT if the templates have virtual methods .. all hell breaks out!

In Felix I handle this as follows: I have a  special shared library which defines ALL the exceptions and this ensures the exception vtable/RTTI pointers are universally unique. I also do not use templates in ways that it would make a difference. I don't use inline methods much. So basically everything is binary code instantiated in specific binary objects.

**Comment by skaller** [comment]
I can try with GCC if you can tell me how to tell cmake to us it .. 🏷

**Comment by skaller** [comment]
I tried this manually:

```
        Start  10: _unit.test.environment.stop_guards.parallel_remove_50k
 10/307 Test  #10: _unit.test.environment.stop_guards.parallel_remove_50k ......................***Exception: Other 30.01 sec
```

and got:
```
~/so-5-5/cmake_build>./test/so_5/environment/stop_guards/parallel_remove_50k/_unit.test.environment.stop_guards.parallel_remove_50k
block of code not finished within 30 seconds. Aborting...
Abort trap: 6
```

**Comment by eao197** [comment]
> Can I ask about use of dynamic cast and exception handling?

Dynamic casts are used widely inside SObjectizer. And I read about problems with RTTI those can arouse when -fvisibility is used. But I can't say that those failed tests do something special.

> I can try with GCC if you can tell me how to tell cmake to us it .. 

I suppose it is something like:
```{.sh}
CMAKE_CXX_COMPILER=g++ CMAKE_C_COMPILER=gcc cmake -DBUILD_TESTS=ON ../dev
```

**Comment by skaller** [comment]
How can I verify its actually using g++?

**Comment by skaller** [comment]
Arghh .. its not: CMakeFiles/CMakeOutput.log

```
The CXX compiler identification is AppleClang, found in "/Users/skaller/so-5-5/cmake_build/CMakeFiles/3.9.0/CompilerIdCXX/a.out"
```

**Comment by skaller** [comment]
ok that should be

```
~/so-5-5/cmake_build>cmake -DCMAKE_CXX_COMPILER=g++ -DCMAKE_C_COMPILER=gcc -DBUILD_TESTS=ON ../dev
```

**Comment by eao197** [comment]
> cmake -DCMAKE_CXX_COMPILER=g++ -DCMAKE_C_COMPILER=gcc

Yes, something like that. I almost don't use CMake :(

**Comment by skaller** [comment]
I don't use make or cmake. Felix bootstraps using fbuild, and after that uses the bootstrap to build its own build tools which it then uses to rebuild a more complete version of itself. Felix `flx` tool does full autobuilding (including for C++) so once you have it you don't ever need a build system again.

Dynamic casts should work ok *inside* the library. They don't work across boundaries even when Eh does. I don't know exactly why but as indicated above, its an ABI, not compiler, problem. The ABI is screwed, end of story. It should have nothing to do with `-fvisibility`, however the weird linker crap used to uniquify RTTI objects might fail if it had the wrong visibility.

Now the thing is, most of al the failures I am seeing are timeouts. Are you using dynamic_casts to identify a particular message kind, which you expect in a mail box, but you're getting the right message there but the cast is failing to see its the desired kind?

**Comment by eao197** [comment]
>  Are you using dynamic_casts to identify a particular message kind, which you expect in a mail box, but you're getting the right message there but the cast is failing to see its the desired kind?

It is hard to say because those are different tests and they work differently. It is necessary to dive into each case and see what is going on. It will do that but it requires time. And now is weekend, my family is not happy when I work on weekends. So I need to take a timeout for a couple of days.

Thanks for your help.

**Comment by skaller** [comment]
BTW: OSX is derived from BSD Unix. So it is using Apple hacked up versions of BSD tools, including dyld, the dynamic linker. Whereas Linux is using hacked up versions of GNU tools. So the compilers are much the same on both systems but the linkers are not.

Hmm .. ok assuming these are g++ built tests .. its hard to know ..this is still failing:

```
 10/307 Test  #10: _unit.test.environment.stop_guards.parallel_remove_50k ......................***Exception: Other 30.01 sec```

**Comment by skaller** [comment]
```
99% tests passed, 1 tests failed out of 307

Total Test time (real) = 174.99 sec

The following tests FAILED:
	 10 - _unit.test.environment.stop_guards.parallel_remove_50k (OTHER_FAULT)
```

Interesting! Now only 1 failure.

**Comment by skaller** [comment]
BTW: I tried to run clang7 but Cmake can't handle it. Don't know why.

Here's a theoretical scenario. In the library you have a base class X *defined in a header*. You get a pointer and dynamic cast to X.

Now in an executable using the library, you make a Y derived from X, and pass the library the pointer. The dynamic cast fails. This is because the executable *also* have a copy of X, and Y is derived from the executable's copy. So the RTTI says its Y is derived from its X, not the X in the library. So the cast in the library can't work, because it's casting to the wrong X.

The solution *for a non  template* is to always define the virtual destructor in a particular translation unit, i.e. a body file, never inline. I am assuming the RTTI/vtables will be generated in the same translation unit as the destructor (that seems sane because its the only possible option). If the linker does "magic merging" across translation units at static link time, this should also work. If the linker puts RTTI/vtables in *every* translation unit where the type is used, it will fail because it will generate the RTTI/vtables in BOTH the executable and library (instead of generating an undefined symbol) UNLESS it also does magic merging when dynamic loading. To make this work the code would have to use one level of indirection, i.e. go through a table filled in when the shared library is loaded.

So what may be happening is the "magic merge" symbol was removed from the shared library symbol table by -fvisibility=hidden. Which would be a compiler bug.

**Comment by eao197** [comment]
I can reproduce some of problems on my FreeBSD VM. Fixes are integrated into the master branch and they are checking by TravisCI right now. I don't know how many of problems are fixed (if any).

**Comment by skaller** [comment]
Template instances are not "dllexport" so they're certain to be hidden ensuring RTTI is duplicated.  I found your dynamic casts, they're in headers which is fine, the problem is duplication of base classes. For a non-template you can fix that, but I have no idea how to do it for templates, in fact I don't think you can: that HAS to be has to be handled by the linker.

Of course you don't need to use system RTTI, but a user space solution, which would be more reliable, would require extensive implementation changes.

**Comment by skaller** [comment]
Status? Using clang? I have a more recent version than the one in High Sierra, but I can't figure out how to use it. Cmake rejects it. I managed to compile all the files with it using my own script but I have no idea which files need to be linked into which libraries/executables. Just curious.

**Comment by eao197** [comment]
TravisCI have successfully compiled the latest version on MacOS with clang (different versions of it): https://travis-ci.org/eao197/so-5-5/builds/480773306

**Comment by skaller** [comment]
Yes, but are you linking to the library dynamically, and building with visibility set to -fvisibility=hidden?

**Comment by eao197** [comment]
As far, as I can tell, yes: https://github.com/eao197/so-5-5/blob/master/dev/so_5/CMakeLists.txt#L110
I could repeat the problem encountered on TravisCI on MacOS on my FreeBSD, made some fixes in code and received clean builds on FreeBSD and on Travis/MacOS.
After that I fixed tag v5.5.24.1.

**Comment by skaller** [comment]
Ok! good job!

FYI, its interesting that with two level namespaces and visibility control, static and dynamic linking no longer work the same. For a single simple library, it probably makes no difference other than some pollution with symbols the client was never supposed to see (but which you just don't document). The impact is minimal in C++ because of namespaces too. The distinction really only starts to cut in with run time plugin loading.

Actually so-5 would really benefit from the user doing that, its the kind of system that would amenable to on demand loading of particular agents. Its really a pity that C and C++ use archaic designs that would make *unloading* very difficult.
